### ISSUE TITLE
Feature Request: Support for WebSocket over Unix Domain Sockets

### ISSUE THREAD

**Issue by zhengfeihe** [boostorg/beast#2918]
Hi Maintainers,

I have found, support for using custom external sockets is not available in beast, as noted in this [issue](https://github.com/boostorg/beast/issues/2055). Specifically, I encountered this limitation when trying to use WebSocket over Unix sockets. This seems to be a missing feature in many C++ WebSocket libraries. There were also issues related to using WebSockets over Unix domain sockets in [websocketpp](https://github.com/zaphoyd/websocketpp/issues/557), but it appears to be unmaintained.

Building WebSocket communication over Unix domain sockets is supported in some packages for other languages. For example, in Python's websocket-client package, you can configure a Unix domain socket when creating client connections: [websocket-client](https://websocket-client.readthedocs.io/en/latest/examples.html#connecting-with-custom-sockets).

Do you think this is a feature that could be added to Beast? If so, would it be acceptable for me to work on this? It may take some time for me to implement.

Best,
Zhengfei

**Comment by ashtum** [comment]
You can easily use a different socket type for the underlying stream of `websocket::stream`:

```C++
using ws_stream_t = beast::websocket::stream<asio::local::stream_protocol::socket>;
```

The [`websocket::stream::next_layer`](https://www.boost.org/doc/libs/develop/libs/beast/doc/html/beast/ref/boost__beast__websocket__stream/next_layer/overload2.html) function gives you access to the underlying stream, allowing you to perform the connect operation. Alternatively, you can move an already connected socket into the websocket stream constructor.

**Comment by vinniefalco** [comment]
We should provide a working example program for this since it keeps coming up.

**Comment by zhengfeihe** [comment]
Thanks for your reply and explanation! I will close the issue.

**Comment by savely-krasovsky** [comment]
I would also add an example with Windows Named Pipe. I know, Windows supports UNIX domain sockets since Windows 10, but named pipes has better Windows Security support (you can run named pipe with own security descriptor).

**Comment by vinniefalco** [comment]
Does asio have a windows named pipe implementation that meets the stream requirements?

**Comment by savely-krasovsky** [comment]
@vinniefalco yes: https://www.boost.org/doc/libs/1_86_0/doc/html/boost_asio/overview/windows/stream_handle.html
I tried it with http client, works perfectly.
